plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetFieldSetterFactory.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+import org.joda.time.DateTimeZone;
+
+import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
+import static io.trino.spi.type.Timestamps.MILLISECONDS_PER_SECOND;
+import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_MICROSECOND;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static java.lang.Math.floorDiv;
+import static java.lang.Math.floorMod;
+import static java.util.Objects.requireNonNull;
+
+public class ParquetFieldSetterFactory
+        extends FieldSetterFactory
+{
+    private final DateTimeZone timeZone;
+
+    public ParquetFieldSetterFactory(DateTimeZone timeZone)
+    {
+        this.timeZone = requireNonNull(timeZone, "time zone is null");
+    }
+
+    @Override
+    public FieldSetter create(SettableStructObjectInspector rowInspector, Object row, StructField field, Type type)
+    {
+        if (type instanceof TimestampType) {
+            return new TimestampFieldSetter(rowInspector, row, field, (TimestampType) type);
+        }
+        return super.create(rowInspector, row, field, type);
+    }
+
+    private class TimestampFieldSetter
+            extends FieldSetter
+    {
+        private final TimestampType type;
+        private final TimestampWritableV2 value = new TimestampWritableV2();
+
+        public TimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, TimestampType type)
+        {
+            super(rowInspector, row, field);
+            this.type = requireNonNull(type, "type is null");
+        }
+
+        @Override
+        public void setField(Block block, int position)
+        {
+            long micros;
+            int nanosOfMicro;
+            if (type.isShort()) {
+                micros = type.getLong(block, position);
+                nanosOfMicro = 0;
+            }
+            else {
+                LongTimestamp longTimestamp = (LongTimestamp) type.getObject(block, position);
+                micros = longTimestamp.getEpochMicros();
+                // TODO: Handle case where we have picos
+                nanosOfMicro = longTimestamp.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND;
+            }
+            int microsOfSecond = floorMod(micros, MICROSECONDS_PER_SECOND);
+            int nanosOfSecond = microsOfSecond * NANOSECONDS_PER_MICROSECOND + nanosOfMicro;
+
+            long seconds = convertLocalEpochSecondsToUtc(floorDiv(micros, MICROSECONDS_PER_SECOND));
+
+            value.set(Timestamp.ofEpochSecond(seconds, nanosOfSecond));
+            rowInspector.setStructFieldData(row, field, value);
+        }
+
+        private long convertLocalEpochSecondsToUtc(long epochSeconds)
+        {
+            long epochMillis = epochSeconds * MILLISECONDS_PER_SECOND;
+            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
+            return epochMillis / MILLISECONDS_PER_SECOND;
```

Comment:
The third commit (`Use inheritance...`) adds back code removed in the second one (`Clean up writing of timestamps in map/array/row to Hive`). This is not a big deal within a PR, but then results in code appearing "out of nowhere" while it's actually being moved.

Perhaps squash second and third commit

Reply:
I've switched the contents a little: now the first commit copies the methods from `HiveWriteUtils`, and the second switches to use inheritance. I think squashing them makes it harder to follow the two distinct changes.